### DISCUSSION TITLE
Recommended way to handle 503 responses from API?

### DISCUSSION THREAD

**Discussion by sbaack** [encode/httpx#1755] [Q&A]
Hi everyone! Thanks for this wonderful tool. I was hoping to get some feedback and suggestions for something I wrote. I've build [a simple tool](https://github.com/sbaack/archive-md-urls) that turns URLs in Markdown files into snapshots from archive.org. [I use httpx's AsyncClient to call the API](https://github.com/sbaack/archive-md-urls/blob/319ec2163a8d4b21b389938a0a99cc20f67854f8/src/archive_md_urls/gather_snapshots.py#L87). The problem is that the API can be slow or unresponsive, in which case it returns a 503 reponse. I've tried to work around this by letting the individual task (each task is a call to the API) wait for five seconds and then retry for a maximum of five times ([here is the code in context](https://github.com/sbaack/archive-md-urls/blob/319ec2163a8d4b21b389938a0a99cc20f67854f8/src/archive_md_urls/gather_snapshots.py#L40)):

```py
tries: int = 0
while tries < 5:
    try:
        response: httpx.Response = await client.get(api_call)
        response.raise_for_status()
        return response.json()
    except httpx.HTTPError:
        tries += 1
        print("Encountered HTTP error, waiting 5 seconds.")
        await asyncio.sleep(5)
        continue
else:
    sys.exit("API unresponsive, try again later.")
```

It kinda works, but I'm really unsure about the 'proper' way of doing this. Would be happy for any feedback and suggestions!

**Comment by lovelydinosaur** [comment]
The good first-pass answer for anything like this would be to figure out what folks are typically doing for `requests`?

(Since we're broadly API compatible)

**Reply by sbaack** [reply]
Thanks for the response! With `requests` it seems to be most common to use `urllib3` for that. See for example [this discussion](https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request). I also found [this article](https://www.peterbe.com/plog/best-practice-with-retries-with-requests) that recommends the same and says that utilizing while-loops similar to how I do is a 'terrible solution' :-/ Given that `urllib3` is synchronous, I assume I can't use it in conjuction with AsyncClient.

Edit: Oh you can [also adjust `requests`' `HTTPAdapter` class for that](https://stackoverflow.com/a/18190415)

**Reply by lovelydinosaur** [reply]
I assume just setting `max_retries` on it's own will just enable connection retries.

*I think* the only support `requests` has for this is through configuring a `urllib3` Retries instance.

We haven't implemented the same set of functionality in `httpx` yet, because it's *kind of* not in the core of `requests` either, and that's been our cut-off for helping figure out what stuff ought to be prioritised for a 1.0 release, and what stuff could wait until later down the road.

**Reply by sbaack** [reply]
Thanks for looking into this. Regardless of whether `httpx` implements this at some point in the future, I still wonder how to best handle `client.get` retries with `AsyncClient` at the moment. Is the while loop I'm using sensible? For aiohttp, there is [aiohttp-retry](https://github.com/inyutin/aiohttp_retry) (haven't tested it). I guess I could try [tenacity](https://github.com/jd/tenacity)

**Reply by StephenBrown2** [reply]
>  I guess I could try tenacity

I think tenacity is the recommended way forward at this current time.

**Reply by ugtar** [reply]
Tenacity is simple but leaves a lot to be desired since it doesn't understand the context of the request. For example, there is no point in retrying 404 errors, but it's generally expected that the client will retry 503 errors. As Tom mentioned, `requests` does have slightly more support for for complex retry setups, but generally `requests` punts on this and relies on urllib3's `Retry` class. It might be prudent for httpx to take the same tack

**Comment by sbaack** [comment]
Following @StephenBrown2's [reply](https://github.com/encode/httpx/discussions/1755#discussioncomment-1027553) I now [implemented this using tenacity](https://github.com/sbaack/archive-md-urls/blob/9b173d81ad241935b71f812045771491654ad378/src/archive_md_urls/gather_snapshots.py#L17). For the record, this is how the updated function currently looks like (I removed the docstring for this preview):

```py
@tenacity.retry(stop=tenacity.stop_after_attempt(5), wait=tenacity.wait_fixed(2))
async def call_api(client: httpx.AsyncClient, api_call: str) -> dict[str, Any]:
    response: httpx.Response = await client.get(api_call)
    response.raise_for_status()
    return response.json()
```
To exit the program when all retries failed (again, code is truncated here):

```py
async with httpx.AsyncClient(timeout=None) as client:
        # Building asyncio task list here, with each task calling call_api() shown above...
        try:
            api_responses: list[dict[str, Any]] = await asyncio.gather(*tasks)
        except tenacity.RetryError:
            sys.exit("API appears unresponsive, please try again later.")
```

Overall, the code is simplified and more readable. I didn't test this a lot, but it seems to work well. If you have any suggestions for improvements, please let me know :-)

Maybe consider adding a section about this to the docs, perhaps with a recommendation for tenacity? And as someone who is not familar with `requests`' more advanced options, it wasn't obvious to me what 'connection retries' means exactly in the section describing [custom transports](https://www.python-httpx.org/advanced/#custom-transports). Some claification would be helpful there as well.

**Reply by stevecj** [reply]
Of course, ideally, HTTPX would use exponential back-off and respect Retry-After headers as urllib3 does. Many services (e.g. Atlassian) expect the client to honor Retry-After.
